pkg/controllers/management/authprovisioningv2/role.go

```diff
@@ -153,6 +157,83 @@ func (h *handler) OnChange(key string, rt *v3.RoleTemplate) (*v3.RoleTemplate, e
 	return rt, nil
 }
 
+func (h *handler) hasClusterAnnotationsAndDeletionTimestamp(annotations map[string]string) (bool, error) {
+	clusterName, cOK := annotations[clusterNameLabel]
+	clusterNamespace, cnOK := annotations[clusterNamespaceLabel]
+	if !cOK || !cnOK {
+		// if this role doesn't have a clustername/namespace label it isn't one of the protected roles, so move on
+		return false, nil
+	}
+	cluster, err := h.clusters.Get(clusterNamespace, clusterName)
+	if err != nil {
+		if apierror.IsNotFound(err) {
+			// when the cluster is not found, we can conclude it has been deleted and safely return
+			return false, nil
+		}
+		return false, err
+	}
+	if cluster == nil {
+		// cluster + err both being nil should not happen - but if it does, safe to delete the role
+		return false, nil
+	}
+	// if the err is nil, we can assume that the cluster is not nil
+	if cluster.DeletionTimestamp == nil {
+		// if the cluster isn't being deleted, it wasn't the source of the delete, so we can remove the role
+		return false, nil
+	}
+	return true, nil
+}
+
+func (h *handler) OnRemoveRole(key string, role *rbacv1.Role) (*rbacv1.Role, error) {
+	shouldEnqueue, err := h.hasClusterAnnotationsAndDeletionTimestamp(role.Annotations)
+	if !shouldEnqueue {
+		return role, err
+	}
+	// Enqueue to ensure perms remain until full delete. Err skips don't re-enqueue, so we do that manually
+	h.roleController.EnqueueAfter(role.Namespace, role.Name, time.Second*5)
+	return role, generic.ErrSkip
+}
+
+func (h *handler) OnRemoveRoleBinding(key string, roleBinding *rbacv1.RoleBinding) (*rbacv1.RoleBinding, error) {
+	shouldEnqueue, err := h.hasClusterAnnotationsAndDeletionTimestamp(roleBinding.Annotations)
+	if !shouldEnqueue {
+		return roleBinding, err
+	}
+	// Enqueue to ensure perms remain until full delete. Err skips don't re-enqueue, so we do that manually
+	h.roleBindingController.EnqueueAfter(roleBinding.Namespace, roleBinding.Name, time.Second*5)
+	return roleBinding, generic.ErrSkip
+}
+
+func (h *handler) OnRemoveClusterRole(key string, role *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+	if role.Annotations == nil {
```

Comment:
You do this check for the ClusterRole/ClusterRoleBinding functions but not for the Role/Rolebinding functions. It would probably be wise to move this check to the `hasClusterAnnotationsAndDeletionTimestamp` function - that way this is consistent across the various functions.

Reply:
That is a very good point